gcc/config/riscv/riscv-insert-vsetvli.cc

```diff
@@ -0,0 +1,1853 @@
+    if (!sew_lmul_ratio_only_p && !other.sew_lmul_ratio_only_p)
+      return vtype_equal_p (other);
+
+    // If the sew_lmul_ratio_only bits are different, then they aren't equal.
+    return false;
+  }
+
+  bool operator!=(const vinfo &Other) const
+  {
+    return !(*this == Other);
+  }
+
+  void
+  operator= (const vinfo &other)
+  {
+    state = other.state;
+    vma = other.vma;
+    vta = other.vta;
+    vsew = other.vsew;
+    vlmul = other.vlmul;
+    all_maskop_p = other.all_maskop_p;
+    sew_lmul_ratio_only_p = other.sew_lmul_ratio_only_p;
+    avl = other.avl;
+    avl_source = other.avl_source;
+  }
+
+  // Calculate the vinfo visible to a block assuming this and other are
+  // both predecessors.
+  vinfo
+  intersect (const vinfo &other) const
+  {
+    // If the new value isn't valid, ignore it.
+    if (!other.valid_p ())
+      return *this;
+
+    // If this value isn't valid, this must be the first predecessor, use it.
+    if (!valid_p ())
+      return other;
+
+    // If either is unknown, the result is unknown.
+    if (unknown_p () || other.unknown_p ())
+      return vinfo::getunknown ();
+
+    // If we have an exact, match return this.
+    if (*this == other)
+      return *this;
+
+    // Not an exact match, but maybe the AVL and VLMAX are the same. If so,
+    // return an SEW/LMUL ratio only value.
+    if (avl_equal_p (other) && vlmax_equal_p (other))
+      {
+	vinfo merge_info = *this;
+	merge_info.sew_lmul_ratio_only_p = true;
+	return merge_info;
+      }
+
+    // otherwise the result is unknown.
+    return vinfo::getunknown ();
+  }
+
+  // Calculate the vinfo visible at the end of the block assuming this
+  // is the predecessor value, and other is change for this block.
+  vinfo
+  merge (const vinfo &other) const
+  {
+    if (!valid_p ())
+      {
+	if (dump_file)
+          fprintf (dump_file, "-------- Can only merge with a valid VSETVLInfo --------\n\n");
+	gcc_unreachable ();
+      }
+
+    // Nothing changed from the predecessor, keep it.
+    if (!other.valid_p ())
+      return *this;
+
+    // If the change is compatible with the input, we won't create a VSETVLI
+    // and should keep the predecessor.
+    if (compatible_p (other))
+      return *this;
+
+    // otherwise just use whatever is in this block.
+    return other;
+  }
+};
+
+struct bb_vinfo
+{
+  // The vinfo that represents the net changes to the VL/VTYPE registers
+  // made by this block. Calculated in Phase 1.
+  vinfo change;
+
+  // The vinfo that represents the VL/VTYPE settings on exit from this
+  // block. Calculated in Phase 2.
+  vinfo exit;
+
+  // The vinfo that represents the VL/VTYPE settings from all predecessor
+  // blocks. Calculated in Phase 2, and used by Phase 3.
+  vinfo pred;
+
+  // Keeps track of whether the block is already in the queue.
+  bool inqueue = false;
+
+  bb_vinfo () {}
+};
+
+std::map<unsigned int, bb_vinfo> m_bb_vinfo_map;
+std::queue<basic_block> m_bb_queue;
```

Comment:
`static` for `m_bb_vinfo_map` and `m_bb_queue`, and rename those two variable, `m_` means member data in general, but apparently that's global variable rather than member data.